lib/data/network/services/app_network_service_impl.dart

```diff
@@ -0,0 +1,65 @@
+import 'dart:convert';
+
+import 'package:basic_auth_challenge/core/exception/exception_utils.dart';
+import 'package:basic_auth_challenge/data/model/dto/login_dto.dart';
+import 'package:basic_auth_challenge/data/model/dto/user_dto.dart';
+import 'package:basic_auth_challenge/data/network/client/dio_client.dart';
+import 'package:basic_auth_challenge/data/network/constants/network_constants.dart';
+import 'package:basic_auth_challenge/data/network/services/app_network_service.dart';
+import 'package:dio/dio.dart';
+
+class AppNetworkServiceImpl extends AppNetworkService {
+  final DioClient client;
+
+  AppNetworkServiceImpl(this.client);
+
+  @override
+  Future<List<UserDto>?> getAllUsers() async {
+    try {
+      final Response response = await client.dio.get(NetworkConstants.allUsers);
+
+      if (response.data != null) {
+        final results = UserResultDto.fromJson(response.data);
+        return results.data;
+      } else {
+        throw ExceptionUtils.handleException(response);
+      }
+    } on Object catch (error) {
+      throw ExceptionUtils.handleException(error);
+    }
+  }
+
+  @override
+  Future<UserDto> getUserDetails(int userId) async {
+    try {
+      final Response response = await client.dio.get(NetworkConstants.allUsers);
+
+      if (response.data != null) {
+        return UserDto.fromJson(response.data['result']);
+      } else {
+        throw ExceptionUtils.handleException(response);
+      }
+    } on Object catch (error) {
+      throw ExceptionUtils.handleException(error);
+    }
+  }
+
+  @override
+  Future<LoginDto> login(String username, password) async {
+    try {
+      String basicAuth =
+          'Basic ${base64.encode(utf8.encode('$username:$password'))}';
+
+      final Response response = await client.dio.post(NetworkConstants.login,
+          options: Options(headers: {"Authorization": basicAuth}));
+
+      if (response.data != null) {
+        return LoginDto.fromJson(response.data['result']);
+      } else {
+        throw ExceptionUtils.handleException(response);
+      }
+    } on Object catch (error) {
+      throw ExceptionUtils.handleException(error);
+    }
+  }
+}
```

Comment:
Nice error handling Faraz but when authentication failed for any reason, there is no indicator for mobile users. 

As it is shown in the snapshot the error already contains the value which might be good to be displayed to the users.

![Screen Shot 1444-02-07 at 11 03 04 AM](https://user-images.githubusercontent.com/54819958/188262200-2d3f2a8c-036a-4733-b6ee-61736b38a05e.png)